_posts/2019-09-20-lambda-to-openfaas.md

```diff
@@ -0,0 +1,282 @@
+```shell
+npm install aws-sdk
+```
+
+#### Use secrets
+We're not moving all of our infrastructure off of AWS, so we'll continue to use DynamoDB. Again, since we want this function to be portable, we'll need to be able to use our AWS Access Key ID and AWS Secret Access Key associated with the same IAM account that our Lambda function was using.
+
+First, you'll need to create a user in AWS with the role that the Lambda function was using. Check the box for "Programmatic access" so that the access keys will be generated. 
+
+Download and save each of the access keys in separate files (with no new lines). We'll read from those files when creating the secrets to ensure that they aren't exposed in any history or logs.
+
+Now, create Kubernetes Secrets for each of those values. You could use the `kubectl create secret ...` command, but to keep with the portability that comes with OpenFaaS, we'll use the OpenFaaS CLI:
+```shell
+$ faas-cli secret create shorturl-dynamo-key --from-file=./access-key-id
+$ faas-cli secret create shorturl-dynamo-secret --from-file=./secret-access-key
+```
+
+To access these secrets in the function, add the following lines:
+```javascript
+let accessKeyId = fs.readFileSync("/var/openfaas/secrets/shorturl-dynamo-key").toString()
+let secretKey = fs.readFileSync("/var/openfaas/secrets/shorturl-dynamo-secret").toString()
+```
+
+Pass them into the AWS SDK configuration so that the DynamoDB `DocumentClient` will have access to the table:
+```javascript
+AWS.config.update({
+    region: 'us-west-1',
+    credentials: {
+        accessKeyId: accessKeyId,
+        secretAccessKey: secretKey
+    }
+});
+```
+Finally, we'll add the secrets to the function's `yml` file so that the function will be given the appropriate permissions to read the secret values
+```yaml
+functions:
+  shortener:
+    lang: node10-express
+    ...
+    secrets:
+      - shorturl-dynamo-key
+      - shorturl-dynamo-secret
+```
+
+> Read more about [Unifying Secrets with OpenFaaS](/blog/unified-secrets)
+
+#### Code Differences
+Below is a picture of what the differences for the "shortener" function look like as seen from a `git diff` (here, using VSCode)
+
+![Function Code Diff](/images/lambda-to-openfaas/code-diff.png)
+
+- We added the `fs` module to be able to read the secrets
+- Removed the `JSON.parse()` call, as the input is already in JSON format, so we are able to read the properties directly
+- Call a new function `configureAWS()` before creating the new `DynamoDB.DocumentClient`
+- Created a new function `configureAWS()` that reads the secrets, and sets the credentials and region for AWS
+- Replaced the AWS specific response object with the template's `context`
+- Set the `context.success` and `context.fail` status codes and values
+
+If you look at the `git diff --stat` output, we actually removed more code than we added!
```

Comment:
With all these changes, is the function still deployable on lambda? Might be worth mentioning

Reply:
Added a note mentioning that because of the changes made, it is no longer able to run on Lambda